### PR TITLE
TimeRange: Add test (and fix code) for infinite timeranges.

### PR DESCRIPTION
**Pull request by C-Lunn**
# Details
A follow up to https://github.com/bbc/rd-apmm-python-lib-mediatimestamp/pull/63 -- I had a vision that I hadn't tested infinite timeranges in the chunker and that was indeed correct.

Done that, found an issue, fixed it.

# Pivotal Story
Story URL:

# Related PRs
https://github.com/bbc/rd-apmm-python-lib-mediatimestamp/pull/63

# Links to external test runs/working deployment
_Where appropriate, if separate to default CI run_

# Submitter PR Checks
_(tick as appropriate)_

- [x] Added bbc/rd-apmm-cloudfit team as a reviewer
- [x] PR completes task/fixes bug
- [x] Tests exercise code appropriately
- [ ] New features and API breaks are flagged in commit messages using magic strings
- [ ] Repo maintainer is notified that a release is required
- [ ] Documentation updated (README, Confluence, Docstrings, API spec, Engineering Guide, etc.)
- [ ] Downstream repos have been checked for potential breaks & fixed as needed
- [ ] APIs/UIs/CLIs updated as required
- [x] PR added to Pivotal story
- [ ] Follow-up stories added to Pivotal
- [ ] Any temporary code/configuration removed (e.g. test deployment environment, temporary commontooling branch)
- [ ] Any pins against pre-releases have been removed

# Reviewer PR Checks
_(tick as appropriate)_

- [ ] PR completes task/fixes bug
- [ ] Tests exercise code appropriately
- [ ] Design makes sense, and fits with our current code base
- [ ] Code is easy to follow
- [ ] PR size is sensible
- [ ] Commit history is sensible and tidy

# Info on Cloudfit PRs
- The checks above are guidelines. They don't all have to be ticked, but they should all have been considered.
- For more details on how to asses PRs, see: https://github.com/bbc/rd-apmm-docs-ways-of-working/blob/master/workflow/PRChecklist.md

### CODE REVIEW

**Review by philipnbbc — COMMENTED** [review]
Not a comment directly related to the changes here, but I wonder whether this method would be simpler if it used some of the methods provided by TimeRange? Particularly `timerange_after()` and `intersect_with()`. That would avoid some of the complexities / edge cases and make it more readable.

For example, you could use
* `bounded_before()` to check it has a start
* `from_start_length(remainder.start, time_duration, ...)` to construct the next (max) chunk timerange to intersect with
* `timerange_after()` to get the remainder timerange
* `intersect_with()` to intersect with the various timeranges
* `is_empty()` to check the remainder is empty

**Comment by C-Lunn** [comment]
> * `bounded_before()` to check it has a start
Is this better than `self.start is not None`?

> * `timerange_after()` to get the remainder timerange
Just as with the current setup, there's an edge case for the first chunk, when no previous chunk exists.  I suppose instead of counting down to a remainer of 0, I could instead maintain a `last_timestamp` that counts up.

> * `intersect_with()` to intersect with the various timeranges
I'm not sure where this would be useful, could you give me an example, please?

> * `is_empty()` to check the remainder is empty
It's possible that the final chunk could have an exclusive end, and so I _think_ that may need an edge case of its own, but it's possible I could have misunderstood the application of it.

**Comment by philipnbbc** [comment]
> > * `bounded_before()` to check it has a start
> 
> Is this better than `self.start is not None`?

It's more descriptive and doesn't require knowledge about what a None start means. It's fine to use `self.start is not None` if self.start is used in places where it should not be None and as it's an internal implementation.

> 
> > * `timerange_after()` to get the remainder timerange
> 
> Just as with the current setup, there's an edge case for the first chunk, when no previous chunk exists. I suppose instead of counting down to a remainer of 0, I could instead maintain a `last_timestamp` that counts up.

My thought would be to start the loop with `remainder = self`. The first chunk is the remainder intersected with the "ideal" chunk constructed from the remainder.start and the chunk duration.

After each yield (e.g. `next_timerange`) the remainder becomes `next_timerange.timerange_after()` intersected with the remainder.

> 
> > * `intersect_with()` to intersect with the various timeranges
> 
> I'm not sure where this would be useful, could you give me an example, please?

If you want the next chunk then create the "ideal" one using `from_start_length()` with inclusive start and exclusive end. Call `intersect_with(remainder_timerange)` and it will be bounded (if it should be shorter) and inclusivities set.

Similarly, to get the remainder you could call `timerange_after()` from the last yielded timerange and then call `intersect_with(remainder_timerange)` and it will be bounded and the inclusivities set.

> 
> > * `is_empty()` to check the remainder is empty
> 
> It's possible that the final chunk could have an exclusive end, and so I _think_ that may need an edge case of its own, but it's possible I could have misunderstood the application of it.

The `intersect_with()` should set the final chunks inclusivity correctly. If it is exclusive and before the end of the ideal chunk length then the chunk will still have an exclusive end. If it is inclusive then the final chunk could end up being just 1 nanosecond.

**Comment by philipnbbc** [comment]
Just to add what I think the core code could look like (untested!):
```
remainder = self
while not remainder.is_empty():
  next_timerange = TimeRange.from_start_length(remainder.start, time_duration, TimeRange.INCLUSIVE_START)
  next_timerange = next_timerange.intersect_with(remainder)

  yield next_timerange
```

**Comment by C-Lunn** [comment]
> Just to add what I think the core code could look like (untested!):
> 
> ```
> remainder = self
> while not remainder.is_empty():
>   next_timerange = TimeRange.from_start_length(remainder.start, time_duration, TimeRange.INCLUSIVE_START)
>   next_timerange = next_timerange.intersect_with(remainder)
> 
>   yield next_timerange
> 
>   remainder = remainder.intersect_with(next_timerange.timerange_after())
> ```

Have updated the method to use this style -- thanks.

File: mediatimestamp/immutable/timerange.py [hunk 1]
```diff
@@ -817,38 +817,26 @@ def into_chunks(
...
-        if last_tr is None:  # There is only one chunk, and it's an exact fit OR shorter than the requested dur.
-            yield self.from_timerange(self)
-        else:
-            if last_tr.end is not None:
-                yield self.from_start_length(
-                    last_tr.end, self.end - last_tr.end,
-                    self.Inclusivity((self.includes_end() << 1) + 1))
-            else:
-                raise ValueError("Something has gone badly wrong, the last chunk had no end value set!")
+        remainder = self  # this is a timerange from the end of the last one to the end of self
```
Comment (philipnbbc) [inline mediatimestamp/immutable/timerange.py:820]: You're missing the check and ValueError for unbounded start / start is None.

File: mediatimestamp/immutable/timerange.py [hunk 2]
```diff
@@ -817,38 +817,26 @@ def into_chunks(
...
+        while not remainder.is_empty():
+            next_timerange = TimeRange.from_start_length(
+                remainder.start,
+                time_duration,
+                self._get_self_inclusivity(remainder, time_duration))
```
Comment (philipnbbc) [inline mediatimestamp/immutable/timerange.py:825]: A couple things that I think could be improved here. The name `_get_self_inclusivity` and it being a member method doesn't look right for something that is specific to this into_chunks method. I would've moved this function into into_chunks. However, I'm not sure a function is really needed.

The start inclusivity case is taken care of by the intersect_with which will demote it to not INCLUDE_START if self doesn't.

The end inclusivity case can be a relatively simple if statement and assignment, e.g.
```
if next_timerange.end == remainder.end and remainder.includes_end():
  next_timerange.inclusivity |= Timerange.INCLUDE_END
```

File: mediatimestamp/immutable/timerange.py [hunk 3]
```diff
@@ -817,38 +817,26 @@ def into_chunks(
...
+            next_timerange = next_timerange.intersect_with(remainder)
+
+            yield next_timerange
+            # If we have just yielded the final chunk, then this will produce an empty timerange.
+            remainder = remainder.intersect_with(next_timerange.timerange_after())
         return
+
+    def _get_self_inclusivity(self, remainder, td):
+        inc = 0b01  # includes_start, excludes end. True for all non-start non-end chunks.
```
Comment (philipnbbc) [inline mediatimestamp/immutable/timerange.py:834]: `inc = TimeRange.INCLUDE_START`

File: mediatimestamp/immutable/timerange.py [hunk 4]
```diff
@@ -817,38 +817,26 @@ def into_chunks(
...
+        if remainder is self:  # this is the first chunk, so set the start bit to to the start inclusivity
+            inc &= 0b10  # clear the second bit
+            inc |= self.includes_start()  # set it to the start inclusivity
```
Comment (philipnbbc) [inline mediatimestamp/immutable/timerange.py:837]: `inc &= ~TimeRange.INCLUDE_START`

File: mediatimestamp/immutable/timerange.py [hunk 5]
```diff
@@ -817,38 +817,26 @@ def into_chunks(
...
+
+        if self.bounded_after() and remainder.start + td >= self.end:  # this is the last chunk
+            inc |= (self.includes_end() << 1)  # set the first bit to the end inclusivity
```
Comment (philipnbbc) [inline mediatimestamp/immutable/timerange.py:840]: `inc |= remainder & TimeRange.INCLUDE_END`

File: mediatimestamp/immutable/timerange.py [hunk 6]
```diff
@@ -817,38 +817,26 @@ def into_chunks(
...
+                time_duration,
+                self._get_self_inclusivity(remainder, time_duration))
+            next_timerange = next_timerange.intersect_with(remainder)
+
+            yield next_timerange
+            # If we have just yielded the final chunk, then this will produce an empty timerange.
+            remainder = remainder.intersect_with(next_timerange.timerange_after())
         return
+
+    def _get_self_inclusivity(self, remainder, td):
```
Comment (philipnbbc) [inline mediatimestamp/immutable/timerange.py:833]: I don't think this method is needed but below would be my comments on what I think should be changed. This follows the existing code style of using the "enum" values for handling inclusivity (I didn't write it btw :-) )

**Review by philipnbbc — COMMENTED** [review]
A couple inline comments of things I think could be simplified

File: mediatimestamp/immutable/timerange.py (hunk 2 above)
Comment (C-Lunn) [reply mediatimestamp/immutable/timerange.py:825]: What about in the first run of the loop when no `next_timerange` exists? This is why I created the function -- I couldn't think of a neater way around it.

Comment (philipnbbc) [reply mediatimestamp/immutable/timerange.py:825]: Good point! I'd forgotten that this TimeRange class is immutable. You could do this instead:
```
if next_timerange.end == remainder.end and remainder.includes_end():
  next_timerange = TimeRange(
    next_timerange.start,
    next_timerange.end,
    next_timerange.inclusivity | TimeRange.INCLUDE_END
  )
```
I think on balance it is better than the separate function as it's more obvious what it is doing.

Comment (C-Lunn) [reply mediatimestamp/immutable/timerange.py:825]: Timeranges are immutable so I don't think it's possible to set the inclusivity _after_ instantiation.

Comment (C-Lunn) [reply mediatimestamp/immutable/timerange.py]: Got around this by creating a new next_timerange

**Review by philipnbbc — APPROVED** [review]
LGTM. Thanks for making the changes!

**Pull request merged**